include/pybind11/embed.h

```diff
@@ -87,29 +87,126 @@ struct embedded_module {
     }
 };
 
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, char** argv, bool add_current_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    char** safe_argv = argv;
+    if (nullptr == argv || argc <= 0) {
+        safe_argv = new char*[1];
+        if (nullptr == safe_argv) return;
+        safe_argv[0] = new char[1];
+        if (nullptr == safe_argv[0]) {
+            delete[] safe_argv;
+            return;
+        }
+        safe_argv[0][0] = '\0';
+        argc = 1;
+    }
+#if PY_MAJOR_VERSION >= 3
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    wchar_t** widened_argv = new wchar_t*[static_cast<unsigned>(argc)];
+    for (int ii = 0; ii < argc; ++ii) {
+#  if PY_MINOR_VERSION >= 5
+        // From Python 3.5 onwards, we're supposed to use Py_DecodeLocale to
+        // generate the wchar_t version of argv.
+        widened_argv[ii] = Py_DecodeLocale(safe_argv[ii], nullptr);
+#    define FREE_WIDENED_ARG(X) PyMem_RawFree(X)
+#  else
+        // Before Python 3.5, we're stuck with mbstowcs, which may or may not
+        // actually work. Mercifully, pyconfig.h provides this define:
+#    ifdef HAVE_BROKEN_MBSTOWCS
+        size_t count = strlen(safe_argv[ii]);
+#    else
+        size_t count = mbstowcs(nullptr, safe_argv[ii], 0);
+#    endif
+        widened_argv[ii] = nullptr;
+        if (count != static_cast<size_t>(-1)) {
+            widened_argv[ii] = new wchar_t[count + 1];
+            mbstowcs(widened_argv[ii], safe_argv[ii], count + 1);
+        }
+#    define FREE_WIDENED_ARG(X) delete[] X
+#  endif
+        if (nullptr == widened_argv[ii]) {
+            // Either we ran out of memory or had a unicode encoding issue.
+            // Free what we've encoded so far and bail.
+            for (--ii; ii >= 0; --ii)
+                FREE_WIDENED_ARG(widened_argv[ii]);
+            return;
+        }
+    }
+
+#  if PY_MINOR_VERSION < 1 || (PY_MINOR_VERSION == 1 && PY_MICRO_VERSION < 3)
+#    define NEED_PYRUN_TO_SANITIZE_PATH 1
+    // don't have SetArgvEx yet
+    PySys_SetArgv(argc, widened_argv);
+#  else
+    PySys_SetArgvEx(argc, widened_argv, add_current_dir_to_path ? 1 : 0);
+#  endif
+
+    // PySys_SetArgv makes new PyUnicode objects so we can clean up this memory
+    if (nullptr != widened_argv) {
+        for (int ii = 0; ii < argc; ++ii)
+            if (nullptr != widened_argv[ii])
+                FREE_WIDENED_ARG(widened_argv[ii]);
+        delete[] widened_argv;
+    }
+#  undef FREE_WIDENED_ARG
+#else
+    // python 2.x
+#  if PY_MINOR_VERSION < 6 || (PY_MINOR_VERSION == 6 && PY_MICRO_VERSION < 6)
+#    define NEED_PYRUN_TO_SANITIZE_PATH 1
```

Comment:
Does this need to be a macro?

Reply:
I'm not sure. My worry was that if I made it a `bool`, a smart optimizer would be able to see that the `PyRun_SimpleString` call is unreachable code.

Reply:
I really, really don't like using macros for flow control.

Reply:
I changed how this code is gated in the rewrite; it's still gated by the `PY_*_VERSION` macros (because I'm not sure how to _not_ do that), but I avoid the `#define`...`#undef`, which I agree was hideous.